mlflow/tracking/artifact_utils.py

```diff
@@ -53,8 +53,15 @@ def _download_artifact_from_uri(artifact_uri, output_path=None):
     :param output_path: The local filesystem path to which to download the artifact. If unspecified,
                         a local output path will be created.
     """
-    artifact_src_dir = posixpath.dirname(artifact_uri)
-    artifact_src_relative_path = posixpath.basename(artifact_uri)
-    artifact_repo = get_artifact_repository(artifact_uri=artifact_src_dir)
-    return artifact_repo.download_artifacts(artifact_path=artifact_src_relative_path,
-                                            dst_path=output_path)
+    parsed_uri = urllib.parse.urlparse(artifact_uri)
+    prefix = ""
+    if parsed_uri.scheme == "file" and not parsed_uri.path.startswith("/"):
```

Comment:
What happens if I use a different URI scheme that is not `file`? Users may plug in their own artifact repo with a different scheme that uses relative paths.

Reply:
I think your previous implementation (up until https://github.com/mlflow/mlflow/pull/1227/commits/9cfb08d533b0ecb7c6a23c73ab6f7d14f97c3ef8, inclusive) worked properly.

According to the documentation of [urlunparse](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlunparse), the re-constructed URI is meant to be semantically equivalent to the URI that was used to produce the `ParseResult`:

```
This may result in a slightly different, but equivalent URL, if the URL that was parsed originally had unnecessary delimiters (for example, a ? with an empty query; the RFC states that these are equivalent).
```

Here's some output from running your previous implementation against various URI types; each tuple is `(original_uri, parsed_root_uri, parsed_artifact_path)`:
```
('file:path/to/artifact', 'file:///path/to', 'artifact')
('file:/path/to/artifact', 'file:///path/to', 'artifact')
('file:///path/to/artifact', 'file:///path/to', 'artifact')
('myscheme:path/to/artifact', 'myscheme:path/to', 'artifact')
('myscheme:/path/to/artifact', 'myscheme:/path/to', 'artifact')
('myscheme:///path/to/artifact', 'myscheme:/path/to', 'artifact')
```

Finally, I wasn't able to find any information about `file` URIs supporting relative paths or `file` URIs of the form `file:path/to/file` here: https://en.wikipedia.org/wiki/File_URI_scheme. Are we sure that these are supported? Regardless, doesn't `urlunparse` seems to handle them properly?

Reply:
Good point on non-file schemes with relative path. I'll remove the scheme == "file" condition.

As for the reconstruction - it is not equivalent if you change relative path to absolute. 
As for whether such paths are valid: look at  https://tools.ietf.org/html/rfc3986#section-3.3
I think that implies that path can be relative if it has no host / authority and path starts with a path segment.